### PR TITLE
Add support for NFS-friendly log protocol.

### PR DESCRIPTION
**Pull request by klizhentas**
Fixes #1549 

* Session events are delivered in continuous
batches in a guaranteed order with every event
and print event ordered from session start.

* Each auth server writes to a separate folder
on disk to make sure that no two processes write
to the same file at a time.

* When retrieving sessions, auth servers fetch
and merge results recorded by each auth server.

* Migrations and compatibility modes are in place
for older clients not aware of the new format,
but compatibility mode is not NFS friendly.

* On disk migrations are launched automatically
during auth server upgrades.

### CODE REVIEW

**Comment by klizhentas** [comment]
retest this please

File: lib/events/doc.go [hunk 1]
```diff
@@ -15,19 +15,77 @@ limitations under the License.
...
+
+On disk file format is designed to be compatible with NFS filesystems and provides
+guarantee that only one auth server writes to the file at a time.
+
+Main Audit Log Format
+=====================
+
+The main log files are saved as:
+
+	/var/lib/teleport/log/<aut-server-id>/<date>.log
```
Comment (klizhentas) [inline lib/events/doc.go]: fix the typo

File: lib/events/doc.go [hunk 2]
```diff
@@ -45,5 +103,129 @@ Common JSON fields
...
+{"file_name":"s1-0.events","type":"events","index":0}
+{"file_name":"s1-0.chunks","type":"chunks","offset":0}
+
+This means that all session events are located in s1-0.events file starting from
+the first event with index 0 and all chunks are located in file s1-0.chunks file
+with the byte offset from the start - 0.
+
+Server a1 will produce the following file:
+
+/var/lib/teleport/a1/s1.index
```
Comment (klizhentas) [inline lib/events/doc.go]: duplicate sentence

File: lib/events/doc.go [hunk 3]
```diff
@@ -45,5 +103,129 @@ Common JSON fields
...
+Server a2 will produce the following file:
+
+/var/lib/teleport/a2/s1.index
+
+With contents:
+
+{"file_name":"s1-3.events","type":"events","index":3}
+{"file_name":"s1-40.chunks","type":"chunks","offset":40}
+
+Events file /var/lib/teleport/a1/s1-4.events will contain:
```
Comment (klizhentas) [inline lib/events/doc.go]: a2

File: lib/events/auditlog.go [hunk 4]
```diff
@@ -230,29 +249,205 @@ func (l *AuditLog) PostSessionChunk(namespace string, sid session.ID, reader io.
...
-	fstream.Seek(int64(offsetBytes), 0)
+	fstream.Seek(int64(offsetBytes)-fileOffset, 0)
 
 	// copy up to maxBytes from the offset position:
 	var buff bytes.Buffer
 	io.Copy(&buff, io.LimitReader(fstream, int64(maxBytes)))
+	return buff.Bytes(), nil
+}
 
+// GetSessionChunk returns a reader which console and web clients request
```
Comment (klizhentas) [inline lib/events/auditlog.go]: delete this code

File: lib/events/auditlog.go [hunk 5]
```diff
@@ -451,6 +852,7 @@ func (l *AuditLog) findInFile(fn string, query url.Values) ([]EventFields, error
 		}
 		for i := range eventFilter {
 			if ef.GetString(EventType) == eventFilter[i] {
+				fmt.Printf("%v has matched %v in filter %v and event %v\n", ef.GetString(EventType), ef, eventFilter[i], eventFilter)
```
Comment (russjones) [inline lib/events/auditlog.go]: log.Debugf?

**Review by russjones — APPROVED** [review]

File: lib/events/auditlog.go (hunk 5 above)
Comment (klizhentas) [reply lib/events/auditlog.go]: yeah, debugging jenkins problems

**Comment by klizhentas** [comment]
retest this

**Comment by klizhentas** [comment]
retest this please

**Pull request merged**